File: rust-hedp/src/customer.rs
//! The input the service validates and renders, ported from internal/customer.
//!
//! Same rules, same limits, same field -> explanation problem map, so a
//! rejected config costs the same on both sides and the benchmark is not
//! quietly comparing different amounts of validation.

use std::collections::BTreeMap;
use std::sync::LazyLock;

use regex::Regex;
use serde::{Deserialize, Serialize};

pub const TIERS: [&str; 4] = ["free", "standard", "premium", "enterprise"];

pub const MAX_FEATURES: usize = 64;
pub const MAX_REPLICAS: i64 = 1000;
pub const MAX_OVERRIDE_DEPTH: usize = 8;
pub const MAX_OVERRIDE_KEYS: usize = 512;

static ID_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[a-z0-9][a-z0-9-]{1,62}$").unwrap());
static REGION_PATTERN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^[a-z]{2}-[a-z]+-[1-9][0-9]?$").unwrap());

#[derive(Debug, Clone, Default, Deserialize, Serialize)]
pub struct Config {
    pub id: String,
    pub tier: String,
    pub region: String,
    #[serde(default)]
    pub features: Vec<String>,
    #[serde(default)]
    pub replicas: i64,
    #[serde(default)]
    pub overrides: BTreeMap<String, serde_json::Value>,
    #[serde(default)]
    pub render_all: bool,
}

impl Config {
    /// validate reports problems as field -> human-readable explanation. An
    /// empty map means the config is renderable.
    pub fn validate(&self, known_features: &[String]) -> BTreeMap<String, String> {
        let mut problems = BTreeMap::new();

        if self.id.is_empty() {
            problems.insert("id".into(), "is required".into());
        } else if !ID_PATTERN.is_match(&self.id) {
            problems.insert(
                "id".into(),
                "must be 2-63 chars of [a-z0-9-] and start with a letter or digit".into(),
            );
        }

        if self.tier.is_empty() {
            problems.insert(
                "tier".into(),
                format!("is required, one of {}", TIERS.join(", ")),
            );
        } else if !TIERS.contains(&self.tier.as_str()) {
            problems.insert(
                "tier".into(),
                format!(
                    "\"{}\" is not a known tier, want one of {}",
                    self.tier,
                    TIERS.join(", ")
                ),
            );
        }

        if self.region.is_empty() {
            problems.insert("region".into(), "is required".into());
        } else if !REGION_PATTERN.is_match(&self.region) {
            problems.insert(
                "region".into(),
                format!(
                    "\"{}\" is not a region identifier, want e.g. eu-west-1",
                    self.region
                ),
            );
        }

        if self.features.len() > MAX_FEATURES {
            problems.insert(
                "features".into(),
                format!(
                    "{} features exceeds the limit of {}",
                    self.features.len(),
                    MAX_FEATURES
                ),
            );
        } else if !known_features.is_empty() {
            let mut seen = std::collections::BTreeSet::new();
            let mut unknown = Vec::new();
            for f in &self.features {
                if !seen.insert(f) {
                    problems.insert("features".into(), format!("\"{f}\" is listed more than once"));
                    continue;
                }
                if !known_features.contains(f) {
                    unknown.push(f.clone());
                }
            }
            if !unknown.is_empty() {
                unknown.sort();
                problems.insert("features".into(), format!("unknown: {}", unknown.join(", ")));
            }
        }

        if self.replicas < 0 || self.replicas > MAX_REPLICAS {
            problems.insert(
                "replicas".into(),
                format!("must be between 0 and {MAX_REPLICAS}"),
            );
        }

        let mut keys = 0usize;
        for (release, override_val) in &self.overrides {
            if release.is_empty() {
                problems.insert("overrides".into(), "release names must not be empty".into());
                break;
            }
            let (depth, k) = shape(override_val, 1);
            keys += k;
            if depth > MAX_OVERRIDE_DEPTH {
                problems.insert(
                    format!("overrides.{release}"),
                    format!("nested {depth} deep, limit is {MAX_OVERRIDE_DEPTH}"),
                );
            }
        }
        if keys > MAX_OVERRIDE_KEYS {
            problems.insert(
                "overrides".into(),
                format!("{keys} total keys exceeds the limit of {MAX_OVERRIDE_KEYS}"),
            );
        }

        problems
    }
}

/// shape returns depth and total key count. Both are bounded before the value
/// reaches the template engine: deep values are cheap to send and expensive to
/// merge.
fn shape(v: &serde_json::Value, depth: usize) -> (usize, usize) {
    let mut max_depth = depth;
    let mut keys = 0;
    match v {
        serde_json::Value::Object(m) => {
            for child in m.values() {
                keys += 1;
                let (d, k) = shape(child, depth + 1);
                keys += k;
                max_depth = max_depth.max(d);
            }
        }
        serde_json::Value::Array(a) => {
            for child in a {
                let (d, k) = shape(child, depth + 1);
                keys += k;
                max_depth = max_depth.max(d);
            }
        }
        _ => {}
    }
    (max_depth, keys)
}

File: golang-hedp/internal/customer/customer.go
// Package customer defines the input the service validates and renders.
//
// Validation runs before any rendering, and it is not decoration: rendering a
// 55 MB chart library to discover that a tier name was misspelled is the
// expensive way to find out. Every check here is one the renderer would
// otherwise pay for in template execution time.
package customer

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

// Config is one customer's configuration. A bulk request carries thousands.
type Config struct {
	ID       string   `json:"id"`
	Tier     string   `json:"tier"`
	Region   string   `json:"region"`
	Features []string `json:"features,omitempty"`
	Replicas int      `json:"replicas,omitempty"`
	// Overrides are per-release value overrides, keyed by release name. They
	// are merged last, over both the blueprint's values and the derived ones.
	Overrides map[string]map[string]any `json:"overrides,omitempty"`
	// RenderAll ignores feature gates and renders the entire library. This is
	// the "does everything still compile" path, and it is orders of magnitude
	// more expensive than a normal request - see the benchmarks.
	RenderAll bool `json:"render_all,omitempty"`
}

// Tiers are the accepted values for Config.Tier.
var Tiers = []string{"free", "standard", "premium", "enterprise"}

var (
	idPattern     = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{1,62}$`)
	regionPattern = regexp.MustCompile(`^[a-z]{2}-[a-z]+-[1-9][0-9]?$`)
)

// Limits bound how much work a single config may ask for. Without them a
// bulk request is an unauthenticated way to allocate arbitrary memory.
const (
	MaxFeatures      = 64
	MaxReplicas      = 1000
	MaxOverrideDepth = 8
	MaxOverrideKeys  = 512
)

// Valid reports problems as field -> human-readable explanation. An empty map
// means the config is renderable.
//
// Returning a map rather than a bare error is what lets a bulk response tell a
// caller which of their 4,000 configs failed and in which field.
func (c Config) Valid(knownFeatures map[string]bool) map[string]string {
	problems := map[string]string{}

	switch {
	case c.ID == "":
		problems["id"] = "is required"
	case !idPattern.MatchString(c.ID):
		problems["id"] = "must be 2-63 chars of [a-z0-9-] and start with a letter or digit"
	}

	if c.Tier == "" {
		problems["tier"] = "is required, one of " + strings.Join(Tiers, ", ")
	} else if !contains(Tiers, c.Tier) {
		problems["tier"] = fmt.Sprintf("%q is not a known tier, want one of %s", c.Tier, strings.Join(Tiers, ", "))
	}

	switch {
	case c.Region == "":
		problems["region"] = "is required"
	case !regionPattern.MatchString(c.Region):
		problems["region"] = fmt.Sprintf("%q is not a region identifier, want e.g. eu-west-1", c.Region)
	}

	if len(c.Features) > MaxFeatures {
		problems["features"] = fmt.Sprintf("%d features exceeds the limit of %d", len(c.Features), MaxFeatures)
	} else if len(knownFeatures) > 0 {
		var unknown []string
		seen := map[string]bool{}
		for _, f := range c.Features {
			if seen[f] {
				problems["features"] = fmt.Sprintf("%q is listed more than once", f)
				continue
			}
			seen[f] = true
			if !knownFeatures[f] {
				unknown = append(unknown, f)
			}
		}
		if len(unknown) > 0 {
			sort.Strings(unknown)
			problems["features"] = fmt.Sprintf("unknown: %s", strings.Join(unknown, ", "))
		}
	}

	if c.Replicas < 0 || c.Replicas > MaxReplicas {
		problems["replicas"] = fmt.Sprintf("must be between 0 and %d", MaxReplicas)
	}

	keys := 0
	for release, override := range c.Overrides {
		if release == "" {
			problems["overrides"] = "release names must not be empty"
			break
		}
		d, k := shape(override, 1)
		keys += k
		if d > MaxOverrideDepth {
			problems["overrides."+release] = fmt.Sprintf("nested %d deep, limit is %d", d, MaxOverrideDepth)
		}
	}
	if keys > MaxOverrideKeys {
		problems["overrides"] = fmt.Sprintf("%d total keys exceeds the limit of %d", keys, MaxOverrideKeys)
	}

	return problems
}

// shape returns the depth and total key count of an arbitrary decoded JSON
// object. Both are bounded before the value is ever handed to the template
// engine: deeply nested values are cheap to send and expensive to coalesce.
func shape(v any, depth int) (maxDepth, keys int) {
	maxDepth = depth
	switch t := v.(type) {
	case map[string]any:
		for _, child := range t {
			keys++
			d, k := shape(child, depth+1)
			keys += k
			if d > maxDepth {
				maxDepth = d
			}
		}
	case []any:
		for _, child := range t {
			d, k := shape(child, depth+1)
			keys += k
			if d > maxDepth {
				maxDepth = d
			}
		}
	}
	return maxDepth, keys
}

func contains(haystack []string, needle string) bool {
	for _, s := range haystack {
		if s == needle {
			return true
		}
	}
	return false
}
